exchange/bitswap/bitswap.go

```diff
@@ -375,6 +373,10 @@ func (bs *Bitswap) ReceiveMessage(ctx context.Context, p peer.ID, incoming bsmsg
 			k := b.Cid()
 			log.Event(ctx, "Bitswap.GetBlockRequest.End", k)
 
+			for _, ses := range bs.SessionsForBlock(k) {
+				ses.receiveBlockFrom(p, b)
```

Comment:
Isn't that a bit racy? That is, what happens if the user adds the block after the call to `blockstore.Get` but before the call to `exchange.GetBlock` ([here](https://github.com/ipfs/go-ipfs/blob/8ab93aaeb220a2ea31b0420479eba4b99cc592d4/blockservice/blockservice.go#L148))?

I'm asking because there's also race condition in this PR when a block is neither in the interest cache (we're interested in too many blocks) nor in the liveWants set (we haven't started fetching it yet) but gets fetched by another session (or manually added by the user). In this case, we'll try to fetch the block from the network anyways even if we already have it.

So, I'm wondering if it's worth adding a notification framework to Blockstores where one can wait for blocks to be added. An alternative is to bitswap with ourself (but that seems kind of hacky).